added reset_idrac and known_good_state clean steps

kpradip001 opened the pull request:
I have added reset_idrac and known_good_state clean steps and it's unit tests in ironic driver, please review it

cdearborn commented on ironic/drivers/modules/drac/raid.py:
```diff
@@ -769,6 +769,35 @@ def delete_configuration(self, task):
 
         return _commit_to_controllers(node, list(controllers))
 
+    @METRICS.timer('DracRAID.reset_idrac')
```
I think these cleaning steps should be in the DracManagement class instead of the RAID class.

kpradip001 replied:
I have added cleaning steps in the DracManagement class in latest commit, can you please review it.

cdearborn replied:
Can you back out the changes to this file?  This file is still showing up as modified.

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/agent.py:
```diff
@@ -813,6 +813,32 @@ def _delete_configuration_final(task, command):
         task.node.raid_config = {}
         task.node.save()
 
+    @METRICS.timer('AgentRAID.reset_idrac')
```
We really shouldn't be making any changes to agent.py.  The cleaning steps that we're adding will be specific to our iDRAC driver.  You would only add the cleaning steps to agent.py if you were adding them for all drivers.

kpradip001 replied:
I have corrected this one in latest commit.

juliakreger replied:
Driver management or bios interfaces please :)

kpradip001 replied:
I have implemented cleaning step in management interface, please check my latest commit.

cdearborn replied:
Can you back out this change?  This is still showing up as a changed file in the PR.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
 
         self.assertEqual({'logical_disks': [expected_logical_disk]},
                          props)
+
```
If we remove the changes to agent.py, then these test cases can be removed.

cdearborn commented on ironic/tests/unit/drivers/modules/test_agent.py:
```diff
@@ -1662,7 +1662,11 @@ def test_get_clean_steps(self, get_steps_mock):
              'priority': 1},
             {'step': 'delete_configuration', 'interface': 'raid',
-             'priority': 2}]
+             'priority': 2},
```
If we remove the changes to agent.py, then these test cases can be removed.

kpradip001 replied:
I have removed this test case, as we don't makes any changes in agent.py.

cdearborn replied:
Can you back out the changes to this file?  It's still showing up as modified.

kpradip001 replied:
Done

kpradip001 commented:
Thanks for correcting me, I will do changes accordingly.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
+        self.assertIsNone(return_value)
+        self.node.refresh()
+
+    @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
```
It looks to me like this test is the same as the one above, but the one above checks the passed parameters.  If this is the case, can you remove this one?

kpradip001 replied:
I have removed this test case, as we don't makes any changes in raid.py.

cdearborn replied:
Can you back out the changes to this file?  This file is still showing up as modified.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_raid.py:
```diff
@@ -1385,3 +1385,68 @@ def test_get_logical_disks(self, mock_list_virtual_disks):
+        self.assertIsNone(return_value)
+        self.node.refresh()
+
+    @mock.patch.object(drac_common, 'get_drac_client', spec_set=True,
```
It looks to me like this test is the same as the one above, but the one above also checks the passed parameters.  If this is the case, can you remove this one?

kpradip001 replied:
I have removed this test case, as we don't makes any changes in raid.py.

kpradip001 commented:
I have addressed all review comments in latest commit, can you please review it?

cdearborn commented on ironic/conf/drac.py:
```diff
@@ -21,7 +21,13 @@
                help=_('Interval (in seconds) between periodic RAID job status '
                       'checks to determine whether the asynchronous RAID '
-                      'configuration was successfully finished or not.'))
+                      'configuration was successfully finished or not.')),
```
Hey Pradip, I don't think we want to be able to set the priority of these cleaning steps in ironic.conf for now.  Can you undo this change?

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
         raise NotImplementedError()
+
+    @METRICS.timer('DracManagement.reset_idrac')
+    @base.clean_step(priority=CONF.drac.clean_priority_reset_idrac)
```
I think the priority for this step should be 0

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
+        client.reset_idrac(force=True, wait=True)
```
I believe you need to tell ironic conductor to resume cleaning after the reset completes like this:
manager_utils.notify_conductor_resume_clean(task)

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
+        client.reset_idrac(force=True, wait=True)
+
+    @METRICS.timer('DracManagement.known_good_state')
+    @base.clean_step(priority=CONF.drac.clean_priority_known_good_state)
```
I think we want this step to have a priority of 0

kpradip001 replied:
Done

cdearborn commented on ironic/drivers/modules/drac/management.py:
```diff
@@ -357,3 +358,32 @@ def get_sensors_data(self, task):
+
+        client = drac_common.get_drac_client(node)
+        client.reset_idrac(force=True, wait=True)
+        client.delete_jobs(job_ids=['JID_CLEARALL_FORCE'])
```
I believe you need to tell ironic conductor to resume cleaning after the reset completes like this:
manager_utils.notify_conductor_resume_clean(task)

kpradip001 replied:
I have added line :  agent_base_vendor._notify_conductor_reume_clean(task)  after delete jobs completes, but while executing cleaning steps, "Failed to tear down from cleaning" exception occurs at this line  -https://github.com/cdearborn/ironic/blob/master/ironic/conductor/manager.py#L1443 and node got stuck into CLEANING state.

cdearborn replied:
I would try calling resume_cleaning the same way that the RAID creation cleaning step does here:
https://github.com/cdearborn/ironic/blob/master/ironic/drivers/modules/drac/raid.py#L903
I'm not sure if the call you are making is doing the same things or not.

kpradip001 commented:
Thanks Chris, I will address mentioned review comments tomorrow.

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -532,3 +532,31 @@ def test_get_sensors_data(self, mock_get_drac_client):
+                force=True, wait=True)
+
+            self.assertIsNone(return_value)
+            self.node.refresh()
```
The node.refresh() line isn't needed above.  It causes the node to be reloaded from the DB, which is only needed before doing asserts on properties of the node.

kpradip001 replied:
Done

cdearborn commented on ironic/tests/unit/drivers/modules/drac/test_management.py:
```diff
@@ -532,3 +532,31 @@ def test_get_sensors_data(self, mock_get_drac_client):
+                job_ids=['JID_CLEARALL_FORCE'])
+
+            self.assertIsNone(return_value)
+            self.node.refresh()
```
The node.refresh() line isn't needed above.  It causes the node to be reloaded from the DB, which is only needed before doing asserts on properties of the node.

kpradip001 replied:
Done

kpradip001 commented:
> Looks great Pradip! I did some poking and found this:
> https://github.com/openstack/ironic/blob/master/ironic/drivers/base.py#L1452
> So it does look like not returning anything is the right thing to do for a synchronous cleaning step.
> I think you can safely remove the node.refresh() calls. Try doing that and running the test cases, if you wouldn't mind. If that all works, then please go ahead and submit the patch to ironic.
> Thanks!

Chris, I have addressed all review comments and created patch to ironic, this is link of patch https://review.opendev.org/#/c/658271/

cdearborn commented:
Thanks Pradip!  I'm going ahead and closing this PR since it has now been submitted to ironic.
